# SARS-COV-2 whole-genome sequence analysis
This repository contains the work that I've done throughout the 3-week [Bioinformatics for Biologists: Analysing and Interpreting Genomics Datasets](https://www.futurelearn.com/courses/bioinformatics-for-biologists-analysing-and-interpreting-genomics-datasets) course, primarily focusing on the implementation of Nextflow for workflow management. 

## ðŸ§¬ Week 1
The first week served as: 
- a revision of file formats used in genomics pipelines (i.e. fasta, fastq, sam, bam, vcf and csv)
- a revision of software management using Conda
- a revision and new insights into DNA sequencing, its history and present applications of short- and long-read sequencing
- an opportunity to strengthen my sequence quality control and mapping abilities
- an opportunity to learn how to identify and characterise genetic variants using Variant Calling

## ðŸ§¬ Week 2
Throughout the second week, I:
- learned about Workflow Management System and nf-core project
- installed Nextflow and created a sample sheet for viralrecon
- ran viralrecon pipeline using Nextflow
- learned how to interpret the viralrecon results and multiqc report

## ðŸ§¬ Week 3
The last week was a great refresher on the best practices in R, throughout which I performed a downstream analysis on the output files from week 2, including data exploration, subsetting and visualisation. 






